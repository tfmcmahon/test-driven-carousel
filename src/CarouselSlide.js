import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const StyledImage = styled.img`
  object-fit: cover;
  width: 100%;
  height: ${({ imgHeight }) =>
    typeof imgHeight === 'number' ? `${imgHeight}px` : imgHeight};
`;

const CarouselSlide = ({
  attribution,
  description,
  imgHeight,
  imgUrl,
  StyledImage,
  ...rest
}) => {
  return (
    <figure {...rest}>
      <StyledImage imgHeight={imgHeight} src={imgUrl} />
      <figcaption>
        <strong>{description}</strong> {attribution}
      </figcaption>
    </figure>
  );
};

CarouselSlide.propTypes = {
  attribution: PropTypes.node,
  description: PropTypes.node.isRequired,
  imgHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  imgUrl: PropTypes.string.isRequired,
  StyledImage: PropTypes.elementType,
};

CarouselSlide.defaultProps = {
  imgHeight: 500,
  StyledImage: StyledImage,
};

export default CarouselSlide;
