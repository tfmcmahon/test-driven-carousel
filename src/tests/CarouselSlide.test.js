import React from 'react';
import { mount, shallow } from 'enzyme';
import CarouselSlide, { StyledImage } from '../CarouselSlide';
import styled from 'styled-components';

describe('CarouselSlide', () => {
  let wrapper;
  const imgUrl = 'https://example.com/image.jpg';
  const description = 'A cool image';
  beforeEach(() => {
    wrapper = shallow(
      <CarouselSlide description={description} imgUrl={imgUrl} />
    );
  });

  it('renders a figure', () => {
    expect(wrapper.type()).toBe('figure');
  });

  it('renders props.StyledImage and a <figcaption> as children', () => {
    expect(wrapper.childAt(0).type()).toEqual(StyledImage);
    expect(wrapper.childAt(1).type()).toBe('figcaption');
  });

  it('passes `imgUrl through to the <img>', () => {
    const img = wrapper.find(StyledImage);
    expect(img.prop('src')).toBe(imgUrl);
  });

  it('uses `description` and `attribution` as the <figcaption>', () => {
    const attribution = 'Gonzo';
    wrapper.setProps({ attribution });
    expect(wrapper.find('figcaption').text()).toBe(
      `${description} ${attribution}`
    );
    expect(wrapper.find('figcaption strong').text()).toBe(description);
  });

  it('passes other props through to the <figure>', () => {
    const style = {};
    const onClick = () => {};
    const className = 'my-carousel-slide';

    wrapper.setProps({ style, onClick, className });

    expect(wrapper.prop('style')).toBe(style);
    expect(wrapper.prop('onClick')).toBe(onClick);
    expect(wrapper.prop('className')).toBe(className);
  });

  it('renders correctly', () => {
    wrapper.setProps({
      attribution: 'Gonzo',
      description: 'A cool image',
    });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('StyledImage', () => {
  let mounted;
  const imgUrl = 'https://example.com/default.jpg';

  beforeEach(() => {
    mounted = mount(<StyledImage src={imgUrl} imgHeight={500} />);
  });

  it('renders an <img> with a given source', () => {
    expect(mounted.containsMatchingElement(<img src={imgUrl} />)).toBe(true);
  });

  it('uses imgHeight as the height style property', () => {
    expect(mounted).toHaveStyleRule('height', '500px');
    mounted.setProps({ imgHeight: 'calc(100vh - 100px)' });
    expect(mounted).toHaveStyleRule('height', 'calc(100vh - 100px)');
  });

  it('allows styles to be overridden', () => {
    const TestImg = styled(CarouselSlide.defaultProps.StyledImage)`
      width: auto;
      height: auto;
      object-fit: fill;
    `;

    mounted = mount(
      <CarouselSlide StyledImage={TestImg} imgUrl={imgUrl} description="Test" />
    );

    expect(mounted.find(TestImg)).toHaveStyleRule('width', 'auto');
    expect(mounted.find(TestImg)).toHaveStyleRule('height', 'auto');
    expect(mounted.find(TestImg)).toHaveStyleRule('object-fit', 'fill');
  });

  it('renders correctly', () => {
    expect(mounted.find('img')).toMatchSnapshot();
  });
});
