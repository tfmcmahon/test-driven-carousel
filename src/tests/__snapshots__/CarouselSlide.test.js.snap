// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselSlide renders correctly 1`] = `
<figure>
  <CarouselSlide__StyledImage
    imgHeight={500}
    src="https://example.com/image.jpg"
  />
  <figcaption>
    <strong>
      A cool image
    </strong>
     
    Gonzo
  </figcaption>
</figure>
`;

exports[`StyledImage renders correctly 1`] = `
.c0 {
  object-fit: cover;
  width: 100%;
  height: 500px;
}

<img
  className="c0"
  src="https://example.com/default.jpg"
/>
`;
